### PR TITLE
[SHIRO-829]: beanPostProcessor and FactoryBean cause spring aop to fail …

### PR DESCRIPTION
**Pull request by xczs666**
When LifecycleBeanPostProcessor and ShiroFilterFactoryBean are defined in the same configuration
class, Realm's dependency aop (@Transactional and cache) is invalidated.

BREAKING CHANGE:
module:shiro-spring class:ShiroFilterFactoryBean

ISSUES CLOSED: #SHIRO-829

Following this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SHIRO) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[SHIRO-XXX] - Fixes bug in SessionManager`,
       where you replace `SHIRO-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install apache-rat:check` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [X] If you have a group of commits related to the same change, please squash your commits into one and force push your branch using `git rebase -i`. 
 
Trivial changes like typos do not require a JIRA issue (javadoc, comments...). 
In this case, just format the pull request title like `(DOC) - Add javadoc in SessionManager`.
 
If this is your first contribution, you have to read the [Contribution Guidelines](https://github.com/apache/shiro/blob/master/CONTRIBUTING.md)

If your pull request is about ~20 lines of code you don't need to sign an [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) 
if you are unsure please ask on the developers list.

To make clear that you license your contribution under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by fpapon** [comment]
@xczs666 it's better to propose this PR on master

**Comment by xczs666** [comment]
> @xczs666 it's better to propose this PR on main and we will cherry-pick on the other branches

Do I need to resubmit a PR based on the main branch?

**Comment by fpapon** [comment]
> > @xczs666 it's better to propose this PR on main and we will cherry-pick on the other branches
> 
> @fpapon Do I need to resubmit a PR based on the main branch?

No it's ok for now ;)

**Comment by xczs666** [comment]
> > > @xczs666 it's better to propose this PR on main and we will cherry-pick on the other branches
> > 
> > 
> > @fpapon Do I need to resubmit a PR based on the main branch?
> 
> No it's ok for now ;)

OK :)

File: support/spring/src/main/java/org/apache/shiro/spring/web/ShiroFilterFactoryBean.java [hunk 1]
```diff
@@ -354,7 +354,7 @@ public void setGlobalFilters(List<String> globalFilters) {
      * @return the application's Shiro Filter instance used to filter incoming web requests.
      * @throws Exception if there is a problem creating the {@code Filter} instance.
      */
-    public Object getObject() throws Exception {
+    public AbstractShiroFilter getObject() throws Exception {
```
Comment (bdemers) [inline support/spring/src/main/java/org/apache/shiro/spring/web/ShiroFilterFactoryBean.java:357]: I like this change, but this is API breaking.

@xczs666 any ideas on how to work around this as you were digging into the problem originally?

Comment (bdemers) [reply support/spring/src/main/java/org/apache/shiro/spring/web/ShiroFilterFactoryBean.java:357]: @xczs666 Thinking about this again, for the 1.x line of Shiro. could we revert lines 357 and 120.  I'm guessing that line 370 is all that's needed?

That said, I wasn't able to reproduce the original issue to test it out either way.  Do you have a simple example project that demos the problem?  From there we should be able write a quick test around the problem.

**Comment by bmarwell** [comment]
We could pull this into 1.9.0 if you like

**Comment by xczs666** [comment]
> We could pull this into 1.9.0 if you like

Please proceed

**Comment by bmarwell** [comment]
Hi @xczs666 we would like to integrate your patch.
However, there is one pending review comment (https://github.com/apache/shiro/pull/316/files#r691462174) and your patch is based on an old version (1.7.0).

Could you kindly rebase your patch and apply Brians suggestion?

Thanks.

**Comment by xczs666** [comment]
> Hi @xczs666 we would like to integrate your patch. However, there is one pending review comment (https://github.com/apache/shiro/pull/316/files#r691462174) and your patch is based on an old version (1.7.0).
> 
> Could you kindly rebase your patch and apply Brians suggestion?
> 
> Thanks.

I can rebase my patch based on the latest version. But I can't solve the problem of returning the signature change. Because the signature of **T getObject()** is declared in the **org.springframework.beans.factory.FactoryBean<T>** interface, to solve this problem need to give **T** a specific type

**Comment by bdemers** [comment]
I didn't have push access to the original repo, but I've reverted the API breaking changes here: https://github.com/xuchenzhi/shiro/pull/1/files
@xczs666 please confirm this will work for you

**Comment by xczs666** [comment]
@bdemers  oh no, in _spring 4.X_ (_springboot 1.X_) the **T** in the class signature of _FactoryBean<T>_ is parsed to get the type instead of calling _getObject()_. In _spring5.X_ (_springboot 2.X_), _getObject()_ was used instead. 
I have a project that simulates this problem: [https://github.com/xczs666/shiro-829](https://github.com/xczs666/shiro-829). You try to modify [https://github.com/xczs666/shiro-829/blob/main/simple-factory-bean-solution-3/src/main/java/simple/factorybean/support/ShiroFilterFactoryBean.java](https://github.com/xczs666/shiro-829/blob/main/simple-factory-bean-solution-3/src/main/java/simple/factorybean/support/ShiroFilterFactoryBean.java), then execute [https://github.com/xczs666/shiro-829/blob/main/simple-factory-bean-solution-3/src/test/java/simple/factorybean/aspect/ServiceAopTest.java](https://github.com/xczs666/shiro-829/blob/main/simple-factory-bean-solution-3/src/test/java/simple/factorybean/aspect/ServiceAopTest.java). You will find that springboot1.X resolves the class signature instead of calling getObject().
I have been busy these two days. Later, I will point out the location of the spring source code for you, as well as the verification demo of different spring versions.

**Comment by bdemers** [comment]
@xczs666 thanks for the info! that really helps!

Can you rebase this onto `1.9.x` or `main` (that should fix the CI issues), and then we can get it merged!

**Comment by fpapon** [comment]
@xczs666 yes, rebase on `main` please, I will cherry-pick on `1.9.x`

**Comment by xczs666** [comment]
@bdemers I tried to fix it rebase on `main`. After modification, a circular dependency error occurs: `ShiroFilterFactoryBean` --_create_--> `AbstractShiroFilter` --_implements_--> `Filter`,  `ShiroFilterFactoryBean` --_depends_--> `Filter`(Map<String, Filter> filters).Means `ShiroFilterFactoryBean` both creates and depends on `Filter`.

Running `ShiroWebConfigurationTest#testMinimalConfiguration` produces exception:

> UnsatisfiedDependencyException: Error creating bean with name 'shiroWebFilterConfiguration': Unsatisfied dependency expressed through field 'filterMap'; nested exception is org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'shiroFilterFactoryBean': Requested bean is currently in creation: Is there an unresolvable circular reference?

> Please clone https://github.com/xczs666/shiro/tree/bugfix/shiro-829 and run **ShiroWebConfigurationTest.testMinimalConfiguration** to reproduce the above error

So I locally removed **filterFactoryBean.setFilters(filterMap)**(line:71) in `org.apache.shiro.spring.web.config.AbstractShiroWebFilterConfiguration`, because `ShiroFilterFactoryBean` will automatically inject Filter Beans through the *postProcessBeforeInitialization(Object, String)* method. 
Although injecting Spring Filter Beans is redundant, accepting this modification will bring about changes in the use of the **ShiroFilterFactoryBean#setFilter** method, which may cause errors in the user's stock code, and you need your help in evaluating the impact and solutions.
Actually in my opinion, in `Spring`, the _setFilters(Map<String, Filter>)_ method should not be provided.

**Comment by bdemers** [comment]
@xczs666 We can merge this into 1.x as is, and I created this issue for v2: [SHIRO-859](https://issues.apache.org/jira/browse/SHIRO-859) to track this

**Comment by xczs666** [comment]
> @xczs666 We can merge this into 1.x as is, and I created this issue for v2: [SHIRO-859](https://issues.apache.org/jira/browse/SHIRO-859) to track this

@bdemers ok then i resubmit a PR based on the `1.9.x` branch?

**Comment by bdemers** [comment]
@xczs666 Thanks! I just switched the target branch to 1.9

**Review by fpapon — APPROVED** [review]
@xczs666 it's ok, @bdemers did the change for the target branch.
Thank you very much for your contribution!

**Pull request merged**